osquery/tables/system/posix/openssl_utils.cpp

```diff
@@ -9,226 +9,392 @@
+
+  auto epoch = std::mktime(&tm);
+  if (epoch == -1) {
+    return boost::none;
   }
-  hex = std::string(hexBytes);
-  OPENSSL_free(hexBytes);
-  return hex;
+
+  return epoch;
 }
 
-bool certificateIsCA(X509* cert) {
-  int ca = X509_check_ca(cert);
-  return (ca > 0);
+} // namespace
+
+boost::optional<std::string> generateCertificateSHA1Digest(X509* cert) {
+  std::array<unsigned char, EVP_MAX_MD_SIZE> binary_digest_buffer{};
+  unsigned int binary_digest_length{};
+
+  if (X509_digest(cert,
+                  EVP_sha1(),
+                  binary_digest_buffer.data(),
+                  &binary_digest_length) == 0) {
+    return boost::none;
+  }
+
+  return convertBinaryDigestToString(binary_digest_buffer.data(),
+                                     binary_digest_length);
 }
 
-bool certificateIsSelfSigned(X509* cert) {
-  bool self_signed = (X509_check_issued(cert, cert) == X509_V_OK);
-  return self_signed;
+void getCertificateAttributes(X509* cert, bool& is_ca, bool& is_self_signed) {
+  is_ca = X509_check_ca(cert) > 0;
+  is_self_signed = (X509_check_issued(cert, cert) == X509_V_OK);
 }
 
-void genCommonName(X509* cert,
-                   std::string& subject,
-                   std::string& common_name,
-                   std::string& issuer) {
-  if (cert == nullptr) {
-    return;
+boost::optional<std::string> getCertificateKeyUsage(X509* cert) {
+  auto key_usage_flags = X509_get_key_usage(cert);
+  if (key_usage_flags == UINT32_MAX) {
+    return boost::none;
   }
 
-  {
-    X509_NAME* issuerName = X509_get_issuer_name(cert);
-    if (issuerName != nullptr) {
-      // Generate the string representation of the issuer.
-      char* issuerBytes = X509_NAME_oneline(issuerName, nullptr, 0);
-      if (issuerBytes != nullptr) {
-        issuer = std::string(issuerBytes);
-        OPENSSL_free(issuerBytes);
-      }
+  std::string output;
+
+  for (const auto& key : kKeyUsageFlagList) {
+    const auto& usage_flag = key.first;
+    if ((key_usage_flags & usage_flag) == 0) {
+      continue;
     }
+
+    if (!output.empty()) {
+      output += ", ";
+    }
+
+    const auto& usage_name = key.second;
+    output += usage_name;
   }
 
-  X509_NAME* subjectName = X509_get_subject_name(cert);
-  if (subjectName == nullptr) {
-    return;
+  return output;
+}
+
+boost::optional<std::string> getCertificateSerialNumber(X509* cert) {
+  auto serial = X509_get_serialNumber(cert);
+  if (serial == nullptr) {
+    return boost::none;
   }
 
-  {
-    // Generate the string representation of the subject.
-    char* subjectBytes = X509_NAME_oneline(subjectName, nullptr, 0);
-    if (subjectBytes != nullptr) {
-      subject = std::string(subjectBytes);
-      OPENSSL_free(subjectBytes);
-    }
+  auto big_num_serial = ASN1_INTEGER_to_BN(serial, nullptr);
+  if (big_num_serial == nullptr) {
+    return boost::none;
   }
 
-  int nid = OBJ_txt2nid("CN");
+  char* string_serial = BN_bn2hex(big_num_serial);
+  OPENSSL_free(big_num_serial);
```

Comment:
This is the wrong free, it should be:
```suggestion
  BN_free(big_num_serial);
```
Asan detects a leak otherwise.